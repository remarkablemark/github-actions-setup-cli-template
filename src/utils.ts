import os from 'os';
import path from 'path';

const architecture = {
  arm: 'arm',
  x32: '386',
  x64: 'amd64',
} as const;

type Arch = keyof typeof architecture;

/**
 * Gets the operating system CPU architecture.
 *
 * @see {@link https://nodejs.org/api/os.html#os_os_arch}
 *
 * @param arch - Arch in [arm, x32, x64...]
 * @returns - Return value in [amd64, 386, arm]
 */
function getArch(arch: Arch) {
  return architecture[arch] || arch;
}

const platform = {
  darwin: 'macOS',
  win32: 'windows',
} as const;

/**
 * Gets a string identifying the operating system platform.
 *
 * @see {@link https://nodejs.org/api/os.html#os_os_platform}
 *
 * @param os - OS in [darwin, linux, win32...]
 * @returns - Return value in [darwin, linux, windows]
 */
function getOS(os: NodeJS.Platform) {
  return platform[os as keyof typeof platform] || os;
}

/**
 * Gets download object.
 *
 * @see {@link https://github.com/cli/cli/releases}
 *
 * @param version - CLI version
 * @returns - URL and binary path
 */
export function getDownloadObject(version: string) {
  const platform = os.platform();
  const arch = os.arch() as Arch;

  const filename = `gh_${version}_${getOS(platform)}_${getArch(arch)}`;
  const extension = platform === 'win32' ? 'zip' : 'tar.gz';

  return {
    binPath: platform === 'win32' ? 'bin' : path.join(filename, 'bin'),
    url: `https://github.com/cli/cli/releases/download/v${version}/${filename}.${extension}`,
  };
}
