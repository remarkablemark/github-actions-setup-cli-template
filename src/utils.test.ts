import os from 'os';

import { getDownloadObject } from './utils';

jest.mock('os');
const mockedOs = jest.mocked(os);

const platforms = ['darwin', 'linux', 'win32'];
const architectures = ['arm', 'x32', 'x64'];

const table = platforms.reduce(
  (testSuites, os) => [
    ...testSuites,
    ...architectures.map((arch) => [os, arch] as [string, string]),
  ],
  [] as [string, string][]
);

describe.each(table)('when OS is %p and arch is %p', (os, arch) => {
  const version = '2.27.0';

  beforeEach(() => {
    jest.resetAllMocks();
    mockedOs.platform.mockReturnValueOnce(os as NodeJS.Platform);
    mockedOs.arch.mockReturnValueOnce(arch);
  });

  it('gets download object', () => {
    expect(getDownloadObject(version)).toMatchSnapshot();
  });
});
