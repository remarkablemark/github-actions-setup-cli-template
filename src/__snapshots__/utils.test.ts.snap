// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when OS is "darwin" and arch is "arm" gets download object 1`] = `
{
  "binPath": "gh_2.27.0_macOS_arm/bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_macOS_arm.tar.gz",
}
`;

exports[`when OS is "darwin" and arch is "x32" gets download object 1`] = `
{
  "binPath": "gh_2.27.0_macOS_386/bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_macOS_386.tar.gz",
}
`;

exports[`when OS is "darwin" and arch is "x64" gets download object 1`] = `
{
  "binPath": "gh_2.27.0_macOS_amd64/bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_macOS_amd64.tar.gz",
}
`;

exports[`when OS is "linux" and arch is "arm" gets download object 1`] = `
{
  "binPath": "gh_2.27.0_linux_arm/bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_linux_arm.tar.gz",
}
`;

exports[`when OS is "linux" and arch is "x32" gets download object 1`] = `
{
  "binPath": "gh_2.27.0_linux_386/bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_linux_386.tar.gz",
}
`;

exports[`when OS is "linux" and arch is "x64" gets download object 1`] = `
{
  "binPath": "gh_2.27.0_linux_amd64/bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_linux_amd64.tar.gz",
}
`;

exports[`when OS is "win32" and arch is "arm" gets download object 1`] = `
{
  "binPath": "bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_windows_arm.zip",
}
`;

exports[`when OS is "win32" and arch is "x32" gets download object 1`] = `
{
  "binPath": "bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_windows_386.zip",
}
`;

exports[`when OS is "win32" and arch is "x64" gets download object 1`] = `
{
  "binPath": "bin",
  "url": "https://github.com/cli/cli/releases/download/v2.27.0/gh_2.27.0_windows_amd64.zip",
}
`;
